app/src/main/java/waiter/EchoServer.java

```diff
@@ -0,0 +1,20 @@
+package waiter;
+
+import waiter.Communicator.Reporter;
+
+import java.io.IOException;
+import java.net.SocketException;
+
+public record EchoServer(Reporter reporter) {
```

Comment:
I'm curious what is the purpose of this `reporter`?

Reply:
hmm. Food for thought: I wonder if the use of interface is making things more difficult to follow. I know that you mentioned testing being the main reason for the interfaces but most of them are very small, with one to two methods at most. We also don't have multiple classes implementing the same interface, I can be wrong. If you change one of the mock's method signature, your test will likely fail so we don't really need interface to enforce that. Part of me thinks we might be better off getting rid of the interfaces. Thoughts? @elijahsullivan @jonmanahan @matteeyao

Reply:
@mrfishball +1 I share your sentiment that there might be too many interfaces.

Reply:
Sorry guys, I'm not able to connect the dots here. But if Jon gets rid of the interfaces, which were being used to mock classes to test another class that uses the class being mocked, would he instead be creating physical test classes to test those classes being fed those mocked classes?

Reply:
In general I don't think testing should be the sole reason for introducing an interface. The purpose of an interface is to:

- make sure the contract doesn't change between multiple implementers and the consumer
- polymorphism in a way that the implementation doesn't matter

In the case where we're testing, for example:
```java
public class Foo {
    ...
    public void doSomething() {
        ...
    }
}

public class FooMock {
    ...
    public void doSomething() {
        ...
    }
}


//test

@Test
void testThatBarWorks() {
   Bar systemUnderTest = new Bar(
       new FooMock,
       ...
   );

   systemUnderTest.doStuff();
   
   assertThatBarWorks
}
```

If we change `FooMock`'s method signature of `doSomething`, I think the test will fail and so my question is do we really need interface so that we can mock a class? There is nothing stopping us from writing a FooMockTwo without an interface.

To @jonmanahan point about making the code more modular. I think writing TDD makes our code more modular. A lot of times we're thinking about how do I go about testing this class and through that process we incorporate dependency injection, single responsibility and other principle to help us write modular code so using interface is like an after the fact thing. I'm curious about the point you make about mock class overriding the original class's method.

Reply:
I think my concern with the interfaces centers more around their use to facilitate the existence of mocks where we might not need them. Perhaps some of these interfaces exist where another would be more appropriate.

FWIW – the example given for an `EchoProtocol` vs an `HTTProtocol` is the kind of interface I like to see.

However, I'm particularly focused on `InputStreamer` and `OutputStreamer`. Neither of which we have tests for from what I can tell, though both perform important behavior.

In the case of the `InputStream` we could test that we call a method on the provided socket (wrapper + interface + mock here) which receives a byte stream and returns a `String`. The inverse with the `OutputStream`.

The `ClientConnection` takes an `InputStreamer` (Reader) and an `OutputStreamer` (Writer), though I wonder if it could just take a socket (again wrapped) and handle reading/writing of `byte streams -> string` and `string -> byte streams` itself. I'm less convinced of the likelihood that our Reader/Writer here will need to change since that is how messages are sent over the wire and more convinced that the socket is what is hard to test here.

Ultimately I think I'm fine keeping these if they make testing the ClientConnection easier, so long as they themselves are tested.

Reply:
@mrfishball That is a true, I think if we change FooMock method signature, the test does indeed fail, but this brings up the question, how would bar in this case handle receiving either a Foo and a FooMock without a common interface between them? If the constructor takes in a FooMock here in the test, then doesn't the constructor in bar have to specify it takes FooMock fooMock and then as a result can't take Foo? (if that makes sense) I could be misunderstanding here so apologies if so and I very much appreciate the analysis here regardless of the outcome!

@elijahsullivan in terms of the `InputStreamer` and `OutputStreamer` I do plan on writing tests for them and unfortunately didn't do so before hand as a result of me misunderstanding the importance in testing their behavior. I do plan on also creating a socket (wrapper mock and interface) for this and the multiple messages ticket as well. As for passing the socket and having `ClientConnection` handle the reading/writing itself, the only issue I see there is that I would either have to new up a reader and writer in the `ClientConnection` constructor, which I don't think/have read is not ideal/good, or I would have to new up the reader/writer in the methods they are used in (ie:read and write in `ClientConnection`) which would cause them to be newuped every call to them, which I don't really like either. Unless I am completely off about the new uping in a constructor, then I think it is possible to pass in just the socket and take care of the rest in `ClientConnection`. Either way I agree, the `InputStreamer` and `OutputStreamer` should be tested if they are staying and I have made note of this to keep track of it and will implement tests for them in the future!

I hope this all made sense and really can't express enough how much I appreciate all of the help and thought provoking discussions not only here but on every PR!

Reply:
+1 @elijahsullivan yea I forgot to mention I do like the Protocol interface in a sense that we know we have a need for it because of echo vs http. I find it very hard to keep track of all the classes we already have on top of the interfaces which makes it difficult to reason about the codebase.

Reply:
Ultimately, I'm not saying we should remove all the interface but rather pop the question of whether we need them or not for the most part. If we do need them are there ways we can organize them in a way that it's obvious that they are interface, for example, class names are usually `nouns`, so maybe we can use adjectives in cases like: 

`Reporter` -> `Reportable`
`Reader` -> `Readable`
`Transporter` -> `Transportable`

Reply:
I completely understand and agree with the fact that the names of the classes and the interfaces are hard to tell apart just by looking at the names, I didn't really think about it from this angle before (not having the IDE to tell you what is an interface and a class) but it makes alot of sense. I like this idea of using adjectives as the interface names alot as it allows one to discern an interface and a class even at a glance. This is a great idea in my opinion, thoughts? @matteeyao @elijahsullivan

Reply:
+1 to the adjective-based interface naming.

Reply:
@jonmanahan re: "I would have to new up the reader/writer in the methods they are used in (ie:read and write in ClientConnection) which would cause them to be newuped every call to them"

I agree, that doesn't seem ideal. Aren't you newing up an`InputStreamReader` and `BufferedReader` on each call to `InputStreamer.readLine()` currently?

Reply:
That is a really good point, Im thinking I should pull out all the new ups into listener then so they are only called once!

Reply:
Addressed in [this commit](https://github.com/jonmanahan/waiter/pull/4/commits/dafffa3b0f3f66e269a98f51dc4d752fc438d5de), thank you!

Reply:
Addressed new Interface naming convention in [this commit](https://github.com/jonmanahan/waiter/pull/4/commits/513240551e9d07c20537e83fd00dbc5039e8cf20), thank you!